### ISSUE TITLE
Crossover points

### ISSUE THREAD

**Issue by philberty** [antoyo/rustc_codegen_gcc#50]
(I was unable to find an email for you so feel free to close this issue)

Hello Antoyo,

Congratulations on all your success with passing libcore tests and upstream alignment with Rustc. 

I thought I would reach out as I am the maintainer for GCC Rust, the front-end project. I believe there is plenty of cross over moments for both projects, as I thought I would share some of my experience with GCC for Rust.

As part of developing the compiler, I have found unit-type support to have some issues in GCC. First, GCC does not seem to understand zero-sized types fully, so my initial implementation for the unit type was by trying void_type_node, which has a special meaning in GCC; this caused crashes with simple assignments. Then, Tom Tromey suggested a zero precision integer to represent it; this seemed to work until we activated optimizations. So now I am using a mixture of zero precision integer and void_type_node, which is working.

For reference, here is a message I posted to GCC: https://gcc.gnu.org/pipermail/gcc-rust/2021-June/000055.html

We have also found that to make sure we debug support correctly, we must prefix tuple field names with __, and GDB understands this convention.

I hope this information is helpful; I think there is a cross over with both projects, and we can help one another with issues like this to get GENERIC to understand rust :).

Brilliant work, hope all is well. Thanks

**Comment by antoyo** [comment]
Yup, that's good to know and we can for sure collaborate for the issues we find.
I'm on vacation currently, so I'll look into these later, but I remember having to use some workarounds for ZSTs.

**Comment by sapir** [comment]
@philberty iirc, in https://github.com/sapir/gcc-rust I used the void type, then just skipped any assignments if they were to ZSTs.

**Comment by philberty** [comment]
> @philberty iirc, in https://github.com/sapir/gcc-rust I used the void type, then just skipped any assignments if they were to ZSTs.

Ah yes I tried this as well but I think I ended up with issues running this test case:

```
fn test(a: ()) -> () {
    a
}

fn main() {
    let a;
    a = test(());
}
```

Where you have ZST parameters in functions.

**Comment by bjorn3** [comment]
ZST parameters in functions should be ignored. They shouldn't even exist in the function signature. This is for example necessary to implement closure->fnptr casts.

**Comment by philberty** [comment]
That's interesting it would simplify things for the backend I might make an issue to do that.

**Comment by philberty** [comment]
We also have an open issue to do with overflow traps. https://github.com/Rust-GCC/gccrs/issues/404 that you might find interesting as well.
